knotx-core/src/main/java/io/knotx/launcher/KnotxStarterVerticle.java

```diff
@@ -16,56 +16,117 @@
  */
 package io.knotx.launcher;
 
+import com.google.common.collect.Lists;
+import io.knotx.Version;
+import io.knotx.launcher.ModuleDescriptor.DeploymentState;
 import io.reactivex.Observable;
-import io.reactivex.ObservableTransformer;
 import io.reactivex.Single;
+import io.vertx.config.ConfigRetrieverOptions;
 import io.vertx.core.DeploymentOptions;
 import io.vertx.core.Future;
 import io.vertx.core.json.JsonObject;
 import io.vertx.core.logging.Logger;
 import io.vertx.core.logging.LoggerFactory;
+import io.vertx.reactivex.config.ConfigRetriever;
 import io.vertx.reactivex.core.AbstractVerticle;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public class KnotxStarterVerticle extends AbstractVerticle {
 
+  private static final String MODULES_ARRAY = "modules";
   private static final String CONFIG_OVERRIDE = "config";
   private static final String MODULE_OPTIONS = "options";
   private static final Logger LOGGER = LoggerFactory.getLogger(KnotxStarterVerticle.class);
+  private List<ModuleDescriptor> deployedModules;
+  private ConfigRetriever configRetriever;
 
   @Override
   public void start(Future<Void> startFuture) {
     printLogo();
-    Observable.fromIterable(config().getJsonArray("modules"))
+
+    if (config().getJsonObject("configRetrieverOptions") == null) {
+      startFuture.fail("Missing 'configRetrieverOptions' in the main config file");
+    }
+
+    configRetriever = ConfigRetriever.create(vertx,
+        new ConfigRetrieverOptions(config().getJsonObject("configRetrieverOptions")));
+
+    configRetriever.listen(conf -> {
+      if (!deployedModules.isEmpty()) {
+        LOGGER.warn("Configuration changed - Re-deploying Knot.x");
+        Observable.fromIterable(deployedModules)
+            .flatMap(item -> vertx.rxUndeploy(item.getDeploymentId()).toObservable())
+            .collect(() -> Lists.newArrayList(),
+                (collector, result) -> collector.add(result))
+            .subscribe(
+                success -> {
+                  LOGGER.warn("Knot.x STOPPED.");
+                  deployVerticles(conf.getNewConfiguration(), null);
```

Comment:
Won't there be NPE in `deployVerticles` in subscribe `onSuccess` / `onError`?

Reply:
good spot. fixed.